### ISSUE TITLE
Problems of detection results of the same object located nearby

### ISSUE THREAD

**Issue by mck0225** [AlexeyAB/darknet#446]
HI AlexeyAB

![test111](https://user-images.githubusercontent.com/26892666/37267458-c3b72d4e-2603-11e8-9198-e03835f5fbce.jpg)

I want to detect the number of plates.
Occasionally, when the same kind of numbers are adjacent, the middle position of the two numbers is detected.
Is there a way to clearly distinguish numbers?

Here is my "cfg".
"tiny-yolo-voc.cfg"
[net]
batch = 64
subdivisions = 32
width = 832
height = 832
channels = 3
momentum = 0.9
decay = 0.0005
angle = 0
saturation = 1.5
exposure = 1.5
hue = .1

learning_rate = 0.0001
max_batches = 100000
policy = steps
steps = 1000,4000
scales = .1, .1

[convolutional]
batch_normalize = 1
filters = 16
size = 3
stride = 1
pad = 1
activation = leaky

[maxpool]
size = 2
stride = 2

[convolutional]
batch_normalize = 1
filters = 32
size = 3
stride = 1
pad = 1
activation = leaky

[maxpool]
size = 2
stride = 2

[convolutional]
batch_normalize = 1
filters = 64
size = 3
stride = 1
pad = 1
activation = leaky

[maxpool]
size = 2
stride = 2

[convolutional]
batch_normalize = 1
filters = 128
size = 3
stride = 1
pad = 1
activation = leaky

[maxpool]
size = 2
stride = 2

[convolutional]
batch_normalize = 1
filters = 256
size = 3
stride = 1
pad = 1
activation = leaky

[maxpool]
size = 2
stride = 2

[convolutional]
batch_normalize = 1
filters = 512
size = 3
stride = 1
pad = 1
activation = leaky

[maxpool]
size = 2
stride = 1

[convolutional]
batch_normalize = 1
filters = 1024
size = 3
stride = 1
pad = 1
activation = leaky

###########

[convolutional]
batch_normalize = 1
size = 3
stride = 1
pad = 1
filters = 1024
activation = linear

[convolutional]
size = 1
stride = 1
pad = 1
filters = 75
activation = linear

[region]
anchors = 0.4528, 1.0.0534, 0.9351, 2.6677, 0.5942, 1.5642, 1.2627, 3.7689, 0.7341, 2.1038
bias_match = 1
classes = 10
coords = 4
num = 5
softmax = 1
jitter = .2
rescore = 1
small_object = 1

object_scale = 5
noobject_scale = 1
class_scale = 1
coord_scale = 1

absolute = 1
thresh = .6
random = 1

**Comment by AlexeyAB** [comment]
The problem is that, 
* image size 4032x3024
* letter size 81x120
* So after image have resized to the network size 832x832, then letter size becomes **16**x33. But the default model of Yolo can well distinguish objects located at a distance of at least **32 pixels** (not 16).


Try to change the model and train again, so this model will can well distinguish objects located at a distance of 16 pixels:

* try to change to `stride=2` in the last `[maxpool]` in cfg-file: https://github.com/AlexeyAB/darknet/blob/15c89e7a714e7e37c13618eace9325a06f0642fc/cfg/tiny-yolo-voc.cfg#L90
* set filters to 256 in the two last convolutional layers: 
    * https://github.com/AlexeyAB/darknet/blob/15c89e7a714e7e37c13618eace9325a06f0642fc/cfg/tiny-yolo-voc.cfg#L94
    * https://github.com/AlexeyAB/darknet/blob/15c89e7a714e7e37c13618eace9325a06f0642fc/cfg/tiny-yolo-voc.cfg#L107

And train again.



----

> anchors = 0.4528, **1.0.0534**, 0.9351, 2.6677, 0.5942, 1.5642, 1.2627, 3.7689, 0.7341, 2.1038

Check your anchors.

**Comment by AlexeyAB** [comment]
@sivagnanamn Problem in the size of final feature map, so if two objects with the same class_id and the same shape (same anchor) are located in the one final activation, then the detection accuracy is greatly reduced, and objects can be coalesced by nms (Non-Maximum suppression).


Yes, Yolo can distinguish very close objects & even overlapping objects, but
* Or parts of objects located at a distance of at least 32 pixels - if objects are big
* Or this detection will be unstable if objects are small, and if both overlaped objects of the same class_id - they can be coalesced by nms

So for ideal detection of small close located objecs, it is required that there is at least 32 pixels after resizing to the network size.
`<input_network_size>  / <final_feature_map_size> = 416 / 13 = 32` 
Or the same `pow(2, <number_of_maxpool with stride==2>) = pow(2, 5) = 32`

So if we remove one <maxpool stride=2> layer, or change to stride=1, then `pow(2,4) = 16`, 

----

Yes, I made a mistake in the answer above, now I fixed it.

So two `[maxpool]` (penultimate and last) layers should have `stride=1`

**Comment by AlexeyAB** [comment]
@mck0225 I fixed my previous answer: 

So both `[maxpool]` (penultimate and last) layers should have `stride=1`

**Comment by mck0225** [comment]
Thank you for your reply.
I'll try learning again.

**Comment by groot-1313** [comment]
@AlexeyAB in order to increase the grid size, you are suggesting the setting of `stride=1` in the last 2 `maxpool` layers. Do the anchor values also need to be modified? Basically, multiplied by 4?

**Comment by AlexeyAB** [comment]
@groot-1313 Yes, you should change anchors:

* or just multiply (width and height of anchors) by 2 

* or recalculate with new final feature map size: 
`darknet.exe detector calc_anchors data/obj.data -num_of_clusters 5 -final_width 26 -final_heigh 26 -show`
